pkg/k8s/policyhelper/policy.go

```diff
@@ -119,3 +156,290 @@ func conflictResolutionSort[T core.Policy](policies []T) {
 		}
 	})
 }
+
+// TODO: remove code above after migration to generic policy handler
+
+// Policy with PolicyTargetReference
+type Policy interface {
+	k8sclient.Object
+	GetStatusConditions() *[]metav1.Condition
+	GetTargetRef() *TargetRef
+}
+
+type PolicyList[P Policy] interface {
+	k8sclient.ObjectList
+	GetItems() []P
+}
+
+type GroupKindSet = utils.Set[GroupKind]
+
+func NewGroupKindSet(l []k8sclient.Object) *GroupKindSet {
+	gks := &GroupKindSet{}
+	for _, obj := range l {
+		gk := ObjGroupKind(obj)
+		gks.Put(gk)
+	}
+	return gks
+}
+
+// A generic handler for common operations on particular policy type
+type PolicyHandler[P Policy] struct {
+	log    gwlog.Logger
+	kinds  *GroupKindSet
+	client PolicyClient[P]
+}
+
+type PolicyHandlerConfig struct {
+	Log            gwlog.Logger
+	Client         k8sclient.Client
+	TargetRefKinds *GroupKindSet
+}
+
+// Creates policy handler for specific policy. T and TL are type and list-type for Policy (struct type, not reference).
+// P and PL are reference types and should derive from T and TL. P and PL do not require explicit declaration. For example:
+//
+//	ph := NewPolicyHandler[IAMAuthPolicy, IAMAuthPolicyList](cfg)
+func NewPolicyHandler[T, TL any, P policyPtr[T], PL policyListPtr[TL, P]](cfg PolicyHandlerConfig) *PolicyHandler[P] {
```

Comment:
Per offline discussion, would be nice if we could infer the policy list type from the policy type

Reply:
That was my initial intent, but couldn't make it work. Go type inference is not that flexible and I'm getting into cyclic dependency issue or type cannot derived and need to be explicitly declared (in this case it's even worse and clumsy)

We can do runtime resolution through type switches, but wont get compiler errors on that. My intent is to let compiler figure-out type miss matches.

We still need at some point explicitly state relationship between Single policy and Plural. Current approach is least amount of code with cost of passing ListType upfront.

Reply:
But I'm open to suggestions, tried few angles here none worked:
- add method to Policy interface to produce list NewList() PolicyList[P] will introduce cyclic dependency

- tried to create a Factory struct that can create new Policy and PolicyList. Go cannot infer type.

```go
type PolicyFactory[P Policy, L PolicyList[P]] interface {
        NewPolicy() P
        NewList() L
}

func NewPolicyHandler[P Policy, L PolicyList[P], F PolicyFactory[P, L]](f F) *PolicyHandler[P] {
        return &PolicyHandler[P]{p: f.NewPolicy()}
}

func main() {
        f := &DummyPolicyFactory{}
        // h := NewPolicyHandler(f) // err: cannot infer P and L
        h := NewPolicyHandler[*DummyPolicy, *DummyPolicyList](f)
}
```

Go cannot infer type from methods of struct, these needs to be explicitly declared.

Reply:
I see, thanks for taking a look into it nonetheless!